packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,196 @@
+   */
+  constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot, { generateNewIds: false })
+
+    let wasRS = false
+    const disposeReactionRS = reaction(
+      () => isRootStore(subtreeRoot),
+      isRS => {
+        if (isRS !== wasRS) {
+          wasRS = isRS
+          if (isRS) {
+            registerRootStore(this.subtreeRootDraft)
+          } else {
+            unregisterRootStore(this.subtreeRootDraft)
+          }
+        }
+      },
+      { fireImmediately: true }
+    )
+
+    const disposeOnPatches = onPatches(subtreeRoot, patches => {
+      if (this.withDraftPatchRecorder) {
+        throw failure("original subtree must not change while 'withDraft' executes")
+      }
+      this.allowWrite(() => {
+        applyPatches(this.subtreeRootDraft, patches)
+      })
+    })
+
+    const { allowWrite, dispose: disposeReadonlyMW } = readonlyMiddleware(this.subtreeRootDraft)
+    this.allowWrite = allowWrite
+
+    this.disposer = () => {
+      disposeReactionRS()
+      disposeOnPatches()
+      disposeReadonlyMW()
+      if (isRootStore(this.subtreeRootDraft)) {
+        unregisterRootStore(this.subtreeRootDraft)
+      }
+      this.disposer = () => {}
+    }
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @typeparam R Return type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is called with a draft copy of `node`. Any changes made to the draft
+   * are applied to the original subtree when `fn` returns `true` or `{ commit: true, ... }`. When
+   * `fn` returns `false` or `{ commit: false, ... }` the changes made to the draft are rejected.
+   * @returns Value of type `R` when `fn` returns an object of type `{ commit: boolean; return: R }`
+   * or `void` when `fn` returns a boolean.
+   */
+  withDraft<T extends object, R = void>(node: T, fn: WithDraftCallback<T, R>): R {
+    assertIsFunction(fn, "fn")
+
+    const { draftNode, applyRecorderChanges } = this.prepareDraftChanges(node)
+
+    let commit = false
+    try {
+      const returnValue = this.allowWrite(() => fn(draftNode))
+      if (typeof returnValue === "boolean") {
+        commit = returnValue
+        return undefined as any
+      } else {
+        commit = returnValue.commit
+        return returnValue.return
+      }
+    } finally {
+      applyRecorderChanges(commit)
+    }
+  }
+
+  /**
+   * Disposes the draft middleware.
+   */
+  dispose(): void {
+    this.disposer()
+  }
+
+  private prepareDraftChanges<T extends object>(
+    node: T
+  ): { draftNode: T; applyRecorderChanges: (commit: boolean) => void } {
+    assertTweakedObject(node, "node")
+
+    const isDraft = !!this.withDraftPatchRecorder
```

Comment:
I assume isDraft means "isNestedWithDraftCall" ?

Reply:
Yes, you're right, I'll change the name.

Reply:
Done